In [1]:
#pip install yfinance
import pandas  as pd #Data manipulation
import numpy as np #Data manipulation

import matplotlib.pyplot as plt # plot graphics
plt.style.use("seaborn")

import yfinance as yf # Yahoo finance
from datetime import datetime

In [2]:
# my portfolio
# "BIMBOA.MX", "HERDEZ.MX", "BOLSAA.MX",
# yf.download(tickers = symbol, start = start, end = end)
symbol = ["GENIUS21.MX", "IVVPESOISHRS.MX", "WALMEX.MX", "AAPL", "AMZN", "FDX","NVDA", "XOM"]
start = "2023-01-02" 
# manyways to say now
end = datetime.now().strftime("%Y-%m-%d")

portfolio_df = yf.download(tickers=symbol, start = start, end = end)
portfolio_df.head()
portfolio_df.info()

[*********************100%%**********************]  8 of 8 completed

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 393 entries, 2023-01-02 to 2024-07-09
Data columns (total 48 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   (Adj Close, AAPL)             380 non-null    float64
 1   (Adj Close, AMZN)             380 non-null    float64
 2   (Adj Close, FDX)              380 non-null    float64
 3   (Adj Close, GENIUS21.MX)      378 non-null    float64
 4   (Adj Close, IVVPESOISHRS.MX)  375 non-null    float64
 5   (Adj Close, NVDA)             380 non-null    float64
 6   (Adj Close, WALMEX.MX)        382 non-null    float64
 7   (Adj Close, XOM)              380 non-null    float64
 8   (Close, AAPL)                 380 non-null    float64
 9   (Close, AMZN)                 380 non-null    float64
 10  (Close, FDX)                  380 non-null    float64
 11  (Close, GENIUS21.MX)          378 non-null    float64
 12  (Close, IVVPESOISHRS.MX)      375 non-null   

In [3]:
# get AAPL stock info
#portfolio_close_df = portfolio_df['Close']
#aapl_df = portfolio_close_df['AAPL']
aapl_df = portfolio_df['Close']['AAPL']

In [4]:
# rest of stocks
amzn_df   = portfolio_df['Close']['AMZN']
fdx_df    = portfolio_df['Close']['FDX']
nvda_df   = portfolio_df['Close']['NVDA']
xom_df    = portfolio_df['Close']['XOM']
gen21_df  = portfolio_df['Close']['GENIUS21.MX']
ivv_df    = portfolio_df['Close']['IVVPESOISHRS.MX']
walmex_df = portfolio_df['Close']['WALMEX.MX']

In [5]:
# defining two data sets, daily and monthly
# https://stackoverflow.com/questions/51604967/return-last-date-and-value-each-month-in-pandas
aapl_df_m = aapl_df.groupby(aapl_df.index.month).tail(1)
aapl_df_m

Date
2023-08-31    187.869995
2023-09-29    171.210007
2023-10-31    170.770004
2023-11-30    189.949997
2023-12-29    192.529999
2024-01-31    184.399994
2024-02-29    180.750000
2024-03-28    171.479996
2024-04-30    170.330002
2024-05-31    192.250000
2024-06-28    210.619995
2024-07-09    228.679993
Name: AAPL, dtype: float64

In [6]:
# rest of stocks
amzn_df_m   = amzn_df.groupby(amzn_df.index.month).tail(1)
fdx_df_m    = fdx_df.groupby(fdx_df.index.month).tail(1)
nvda_df_m   = nvda_df.groupby(nvda_df.index.month).tail(1)
xom_df_m    = xom_df.groupby(xom_df.index.month).tail(1)
gen21_df_m  = gen21_df.groupby(gen21_df.index.month).tail(1)
ivv_df_m    = ivv_df.groupby(ivv_df.index.month).tail(1)
walmex_df_m = walmex_df.groupby(walmex_df.index.month).tail(1)

In [7]:
# interactive plots cufflinks 
# import it or install it
try:
    import cufflinks as cf
except:
    %pip install cufflinks
    import cufflinks as cf

In [8]:
cf.set_config_file(offline = True)
aapl_df.iplot()

In [9]:
# creating US df with close values
us_portfolio_df = pd.DataFrame()
us_portfolio_df['AAPL'] = aapl_df.dropna()
us_portfolio_df['AMZN'] = amzn_df.dropna()
us_portfolio_df['FDX']  = fdx_df.dropna()
us_portfolio_df['NVDA'] = nvda_df.dropna()
us_portfolio_df['XOM']  = xom_df.dropna()

# https://numpy.org/doc/stable/reference/generated/numpy.log.html
#us_portfolio_df["AAPL_Returns"] = np.log(aapl_df.div(aapl_df.shift(1)))
us_portfolio_df["AAPL_Returns"] = aapl_df.pct_change()
#us_portfolio_df["AMZN_Returns"] = np.log(amzn_df.div(amzn_df.shift(1)))
us_portfolio_df["AMZN_Returns"] = amzn_df.pct_change()
#us_portfolio_df["FDX_Returns"] = np.log(fdx_df.div(fdx_df.shift(1)))
us_portfolio_df["FDX_Returns"] = fdx_df.pct_change()
#us_portfolio_df["NVDA_Returns"] = np.log(nvda_df.div(nvda_df.shift(1)))
us_portfolio_df["NVDA_Returns"] = nvda_df.pct_change()
#us_portfolio_df["XOM_Returns"] = np.log(xom_df.div(xom_df.shift(1)))
us_portfolio_df["XOM_Returns"] = xom_df.pct_change()

us_portfolio_df

AAPL        AMZN         FDX        NVDA         XOM  \
Date                                                                     
2023-01-03  125.070000   85.820000  177.270004   14.315000  106.510002   
2023-01-04  126.360001   85.139999  181.449997   14.749000  106.820000   
2023-01-05  125.019997   83.120003  181.279999   14.265000  109.209999   
2023-01-06  129.619995   86.080002  185.770004   14.859000  110.529999   
2023-01-09  130.149994   87.360001  189.740005   15.628000  108.470001   
...                ...         ...         ...         ...         ...   
2024-07-02  220.270004  200.000000  298.220001  122.669998  114.180000   
2024-07-03  221.550003  197.589996  293.709991  128.279999  114.760002   
2024-07-05  226.339996  200.000000  294.000000  125.830002  113.370003   
2024-07-08  227.820007  199.289993  295.529999  128.199997  112.180000   
2024-07-09  228.679993  199.339996  292.730011  131.380005  110.940002   

            AAPL_Returns  AMZN_Returns  FDX_Returns  NVDA_Returns  XOM_Returns  
Date                                                                            
2023-01-03           NaN           NaN          NaN           NaN          NaN  
2023-01-04      0.010314     -0.007924     0.023580      0.030318     0.002911  
2023-01-05     -0.010605     -0.023726    -0.000937     -0.032816     0.022374  
2023-01-06      0.036794      0.035611     0.024768      0.041640     0.012087  
2023-01-09      0.004089      0.014870     0.021371      0.051753    -0.018637  
...                  ...           ...          ...           ...          ...  
2024-07-02      0.016240      0.014199    -0.001540     -0.013113    -0.006785  
2024-07-03      0.005811     -0.012050    -0.015123      0.045732     0.005080  
2024-07-05      0.021620      0.012197     0.000987     -0.019099    -0.012112  
2024-07-08      0.006539     -0.003550     0.005204      0.018835    -0.010497  
2024-07-09      0.003775      0.000251    -0.009474      0.024805    -0.011054  

[380 rows x 10 columns]

In [10]:
# create US dayly df difference with values
us_portfolio_diff_df = pd.DataFrame()
ticks_us = ['AAPL', 'AMZN', 'FDX', 'NVDA', 'XOM']

for t in ticks_us:
    # transform into differences dt
    tmp_l = []
    j = 0
    for i in us_portfolio_df[t].index:
        #print(t, i, j)
        if j > 0:
            # look for consecutive cells
            #print(us_portfolio_df[t].iloc[j-1] - us_portfolio_df[t].iloc[j])
            tmp_l.append(us_portfolio_df[t].iloc[j] - us_portfolio_df[t].iloc[j-1])
        j = j + 1
    us_portfolio_diff_df = pd.concat([us_portfolio_diff_df, pd.DataFrame(tmp_l, columns=[t])], axis=1)

#tmp_portfolio_diff_df = pd.DataFrame(tmp_l, columns=[t])
#us_portfolio_diff_df = pd.concat([us_portfolio_diff_df, tmp_portfolio_diff_df], ignore_index=True, sort=False)

us_portfolio_diff_df

AAPL      AMZN       FDX      NVDA       XOM
0    1.290001 -0.680000  4.179993  0.434000  0.309998
1   -1.340004 -2.019997 -0.169998 -0.483999  2.389999
2    4.599998  2.959999  4.490005  0.594000  1.320000
3    0.529999  1.279999  3.970001  0.769000 -2.059998
4    0.580002  2.510002 -1.000000  0.281000  1.619995
..        ...       ...       ...       ...       ...
374  3.520004  2.800003 -0.459991 -1.630005 -0.779999
375  1.279999 -2.410004 -4.510010  5.610001  0.580002
376  4.789993  2.410004  0.290009 -2.449997 -1.389999
377  1.480011 -0.710007  1.529999  2.369995 -1.190002
378  0.859985  0.050003 -2.799988  3.180008 -1.239998

[379 rows x 5 columns]

In [11]:
# plot us portfolio
cf.set_config_file(offline = True)
us_portfolio_diff_df.iplot()

In [12]:
# plot us portfolio
cf.set_config_file(offline = True)
us_portfolio_df.iplot()

In [13]:
# create US monthly df difference with values
us_portfolio_m_df = pd.DataFrame()
us_portfolio_m_diff_df = pd.DataFrame()
ticks_us = ['AAPL', 'AMZN', 'FDX', 'NVDA', 'XOM']

us_portfolio_m_df['AAPL'] = aapl_df_m.dropna()
us_portfolio_m_df['AMZN'] = amzn_df_m.dropna()
us_portfolio_m_df['FDX']  = fdx_df_m.dropna()
us_portfolio_m_df['NVDA'] = nvda_df_m.dropna()
us_portfolio_m_df['XOM']  = xom_df_m.dropna()

for t in ticks_us:
    # transform into differences dt
    tmp_l = []
    j = 0
    for i in us_portfolio_m_df[t].index:
        #print(t, i, j)
        if j > 0:
            # look for consecutive cells
            #print(us_portfolio_df[t].iloc[j-1] - us_portfolio_df[t].iloc[j])
            tmp_l.append(us_portfolio_m_df[t].iloc[j] - us_portfolio_m_df[t].iloc[j-1])
        j = j + 1
    us_portfolio_m_diff_df = pd.concat([us_portfolio_m_diff_df, pd.DataFrame(tmp_l, columns=[t])], axis=1)

#tmp_portfolio_diff_df = pd.DataFrame(tmp_l, columns=[t])
#us_portfolio_diff_df = pd.concat([us_portfolio_diff_df, tmp_portfolio_diff_df], ignore_index=True, sort=False)

us_portfolio_m_diff_df

AAPL       AMZN        FDX       NVDA        XOM
0  -16.659988 -10.889992   3.900024  -5.855999   6.389999
1   -0.440002   5.969994 -24.820007  -2.719002 -11.730003
2   19.179993  13.000000  18.729980   5.990002  -3.110001
3    2.580002   5.850006  -5.859985   2.751999  -2.759995
4   -8.130005   3.259995 -11.680008  12.005001   2.829994
5   -3.649994  21.559998   7.680008  17.584999   1.709999
6   -9.270004   3.620010  40.769989  11.244003  11.720001
7   -1.149994  -5.380005 -27.959991  -3.954002   2.029999
8   21.919998   1.440002  -7.819992  23.231003  -1.009995
9   18.369995  16.809998  45.879990  13.906998  -2.139999
10  18.059998   6.089996  -7.109985   7.840004  -4.180000

In [14]:
# creating MX portfolio
mx_portfolio_df = pd.DataFrame()
mx_portfolio_df['GENIUS21.MX'] = gen21_df.dropna()
mx_portfolio_df['IVVPESOISHRS.MX'] = ivv_df.dropna()
mx_portfolio_df['WALMEX.MX'] = walmex_df.dropna()
# calc Returns
# https://numpy.org/doc/stable/reference/generated/numpy.log.html
#mx_portfolio_df["GENIUS21.MX_Returns"] = np.log(gen21_df.div(gen21_df.shift(1)))
mx_portfolio_df["GENIUS21.MX_Returns"] = gen21_df.pct_change()
#mx_portfolio_df["IVVPESOISHRS.MX_Returns"] = np.log(ivv_df.div(ivv_df.shift(1)))
mx_portfolio_df["IVVPESOISHRS.MX_Returns"] = ivv_df.pct_change()
#mx_portfolio_df["WALMEX.MX_Returns"] = np.log(walmex_df.div(walmex_df.shift(1)))
mx_portfolio_df["WALMEX.MX_Returns"] = walmex_df.pct_change()

mx_portfolio_df

GENIUS21.MX  IVVPESOISHRS.MX  WALMEX.MX  GENIUS21.MX_Returns  \
Date                                                                       
2023-01-02    37.230000              NaN  69.150002                  NaN   
2023-01-03    36.799999        70.349998  69.250000            -0.011550   
2023-01-04    37.450001        70.769997  70.489998             0.017663   
2023-01-05    36.860001        70.339996  70.410004            -0.015754   
2023-01-06    37.730000        72.000000  72.449997             0.023603   
...                 ...              ...        ...                  ...   
2024-07-03    88.199997       107.000000  61.560001             0.006390   
2024-07-04    87.720001       107.050003  61.549999            -0.005442   
2024-07-05    88.949997       105.800003  61.619999             0.014022   
2024-07-08    88.680000       105.879997  61.830002            -0.003035   
2024-07-09    88.459999       106.139999  64.059998            -0.002481   

            IVVPESOISHRS.MX_Returns  WALMEX.MX_Returns  
Date                                                    
2023-01-02                      NaN                NaN  
2023-01-03                      NaN           0.001446  
2023-01-04                 0.005970           0.017906  
2023-01-05                -0.006076          -0.001135  
2023-01-06                 0.023600           0.028973  
...                             ...                ...  
2024-07-03                 0.025002          -0.002269  
2024-07-04                 0.000467          -0.000162  
2024-07-05                -0.011677           0.001137  
2024-07-08                 0.000756           0.003408  
2024-07-09                 0.002456           0.036067  

[378 rows x 6 columns]

In [15]:
# create MX dayly df difference with values
mx_portfolio_diff_df = pd.DataFrame()
ticks_mx = ['GENIUS21.MX', 'IVVPESOISHRS.MX', 'WALMEX.MX']

for t in ticks_mx:
    # transform into differences dt
    j = 0
    tmp_l = []
    for i in mx_portfolio_df[t].index:
        #print(t, i, j)
        if j > 0:
            # look for consecutive cells
            #print(us_portfolio_df[t].iloc[j-1] - us_portfolio_df[t].iloc[j])
            tmp_l.append(mx_portfolio_df[t].iloc[j] - mx_portfolio_df[t].iloc[j-1])
        j = j + 1
    mx_portfolio_diff_df = pd.concat([mx_portfolio_diff_df, pd.DataFrame(tmp_l, columns=[t])], axis=1)

mx_portfolio_diff_df

GENIUS21.MX  IVVPESOISHRS.MX  WALMEX.MX
0      -0.430000              NaN   0.099998
1       0.650002         0.419998   1.239998
2      -0.590000        -0.430000  -0.079994
3       0.869999         1.660004   2.039993
4       0.180000         0.150002  -0.079994
..           ...              ...        ...
372     0.559998         2.610001  -0.139999
373    -0.479996         0.050003  -0.010002
374     1.229996        -1.250000   0.070000
375    -0.269997         0.079994   0.210003
376    -0.220001         0.260002   2.229996

[377 rows x 3 columns]

In [16]:
# plot mx portfolio
cf.set_config_file(offline = True)
mx_portfolio_diff_df.iplot()

In [17]:
# plot mx portfolio
cf.set_config_file(offline = True)
mx_portfolio_df.iplot()

In [18]:
# create MX monthly df difference with values
# = aapl_df.groupby(aapl_df.index.month).tail(1)
#   = amzn_df.groupby(amzn_df.index.month).tail(1)
#    = fdx_df.groupby(fdx_df.index.month).tail(1)
#   = nvda_df.groupby(nvda_df.index.month).tail(1)
#    = xom_df.groupby(xom_df.index.month).tail(1)
#  = gen21_df.groupby(gen21_df.index.month).tail(1)
#    = ivv_df.groupby(ivv_df.index.month).tail(1)
# = walmex_df.groupby(walmex_df.index.month).tail(1)
mx_portfolio_m_df = pd.DataFrame()
mx_portfolio_m_diff_df = pd.DataFrame()
ticks_mx = ['GENIUS21.MX', 'IVVPESOISHRS.MX', 'WALMEX.MX']

mx_portfolio_m_df['GENIUS21.MX'] = gen21_df_m.dropna()
mx_portfolio_m_df['IVVPESOISHRS.MX'] = ivv_df_m.dropna()
mx_portfolio_m_df['WALMEX.MX'] = walmex_df_m.dropna()

for t in ticks_mx:
    # transform into differences dt
    tmp_l = []
    j = 0
    for i in mx_portfolio_m_df[t].index:
        #print(t, i, j)
        if j > 0:
            # look for consecutive cells
            #print(us_portfolio_df[t].iloc[j-1] - us_portfolio_df[t].iloc[j])
            tmp_l.append(mx_portfolio_m_df[t].iloc[j] - mx_portfolio_m_df[t].iloc[j-1])
        j = j + 1
    mx_portfolio_m_diff_df = pd.concat([mx_portfolio_m_diff_df, pd.DataFrame(tmp_l, columns=[t])], axis=1)

#tmp_portfolio_diff_df = pd.DataFrame(tmp_l, columns=[t])
#us_portfolio_diff_df = pd.concat([us_portfolio_diff_df, tmp_portfolio_diff_df], ignore_index=True, sort=False)

mx_portfolio_m_diff_df

GENIUS21.MX  IVVPESOISHRS.MX  WALMEX.MX
0    -2.969997        -4.080002  -1.440002
1     1.480000        -1.650002  -1.180000
2     4.189999         7.400002   3.779999
3     0.720001         4.340004   3.010002
4     6.539997         1.469994  -0.300003
5     8.520000         4.920006  -3.009995
6    -1.419998         0.059998  -4.230003
7     6.949997         2.839996  -0.099998
8    11.580002         5.480003  -1.259998
9     1.820000         2.389999   1.569996

In [19]:
# create us analysis portfolio dayly info
index_l = ['Day_outcome', 'Month_outcome', 'Year_outcome', 'Year_var', 'Cov_year_mkt']
us_portfolio_analysis_df = pd.DataFrame(index = index_l)
#print(us_portfolio_df['AAPL'].mean())
us_portfolio_analysis_df['AAPL'] = us_portfolio_diff_df['AAPL'].mean()
us_portfolio_analysis_df['AMZN'] = us_portfolio_diff_df['AMZN'].mean()
us_portfolio_analysis_df['FDX'] = us_portfolio_diff_df['FDX'].mean()
us_portfolio_analysis_df['NVDA'] = us_portfolio_diff_df['NVDA'].mean()
us_portfolio_analysis_df['XOM']  = us_portfolio_diff_df['XOM'].mean()
us_portfolio_analysis_df


AAPL      AMZN       FDX      NVDA       XOM
Day_outcome    0.273377  0.299525  0.304644  0.308879  0.011689
Month_outcome  0.273377  0.299525  0.304644  0.308879  0.011689
Year_outcome   0.273377  0.299525  0.304644  0.308879  0.011689
Year_var       0.273377  0.299525  0.304644  0.308879  0.011689
Cov_year_mkt   0.273377  0.299525  0.304644  0.308879  0.011689

In [20]:
# us portfolio monthly info
#aapl_df_m = aapl_df.groupby(aapl_df.index.month).tail(1)
#print(aapl_df_m.mean())
us_portfolio_analysis_df['AAPL']['Month_outcome'] = us_portfolio_m_diff_df['AAPL'].mean()
us_portfolio_analysis_df['AAPL']['Year_outcome']  = us_portfolio_m_diff_df['AAPL'].mean()*12
us_portfolio_analysis_df['AAPL']['Year_var']      = us_portfolio_df["AAPL_Returns"].var()
#us_portfolio_analysis_df['AAPL']['Cov_year_mkt']  = us_portfolio_df["AAPL_Returns"].cov()
us_portfolio_analysis_df['AAPL']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["AAPL_Returns"].var())
#amzn_df_m   = amzn_df.groupby(amzn_df.index.month).tail(1)
us_portfolio_analysis_df['AMZN']['Month_outcome'] = us_portfolio_m_diff_df['AMZN'].mean()
us_portfolio_analysis_df['AMZN']['Year_outcome']  = us_portfolio_m_diff_df['AMZN'].mean()*12
us_portfolio_analysis_df['AMZN']['Year_var']      = us_portfolio_df["AMZN_Returns"].var()
#us_portfolio_analysis_df['AMZN']['Cov_year_mkt']  = us_portfolio_df["AMZN_Returns"].cov()
us_portfolio_analysis_df['AMZN']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["AMZN_Returns"].var())
#fdx_df_m   = fdx_df.groupby(amzn_df.index.month).tail(1)
us_portfolio_analysis_df['FDX']['Month_outcome'] = us_portfolio_m_diff_df['FDX'].mean()
us_portfolio_analysis_df['FDX']['Year_outcome']  = us_portfolio_m_diff_df['FDX'].mean()*12
us_portfolio_analysis_df['FDX']['Year_var']      = us_portfolio_df["FDX_Returns"].var()
#us_portfolio_analysis_df['FDX']['Cov_year_mkt']  = us_portfolio_df["FDX_Returns"].cov()
us_portfolio_analysis_df['FDX']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["FDX_Returns"].var())
#nvda_df_m   = nvda_df.groupby(nvda_df.index.month).tail(1)
us_portfolio_analysis_df['NVDA']['Month_outcome'] = us_portfolio_m_diff_df['NVDA'].mean()
us_portfolio_analysis_df['NVDA']['Year_outcome']  = us_portfolio_m_diff_df['NVDA'].mean()*12
us_portfolio_analysis_df['NVDA']['Year_var']      = us_portfolio_df["NVDA_Returns"].var()
#us_portfolio_analysis_df['NVDA']['Cov_year_mkt']  = us_portfolio_df["NVDA_Returns"].cov()
us_portfolio_analysis_df['NVDA']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["NVDA_Returns"].var())
#xom_df_m    = xom_df.groupby(xom_df.index.month).tail(1)
us_portfolio_analysis_df['XOM']['Month_outcome']  = us_portfolio_m_diff_df['XOM'].mean()
us_portfolio_analysis_df['XOM']['Year_outcome']   = us_portfolio_m_diff_df['XOM'].mean()*12
us_portfolio_analysis_df['XOM']['Year_var']       = us_portfolio_df["XOM_Returns"].var()
#us_portfolio_analysis_df['XOM']['Cov_year_mkt']   = us_portfolio_df["XOM_Returns"].cov()
us_portfolio_analysis_df['XOM']['Cov_year_mkt']   = np.sqrt(us_portfolio_df["XOM_Returns"].var())

us_portfolio_analysis_df

AAPL       AMZN        FDX       NVDA       XOM
Day_outcome     0.273377   0.299525   0.304644   0.308879  0.011689
Month_outcome   3.710000   5.575455   2.882729   7.456819 -0.022727
Year_outcome   44.519997  66.905457  34.592751  89.481824 -0.272727
Year_var        0.000186   0.000367   0.000319   0.000975  0.000208
Cov_year_mkt    0.013649   0.019159   0.017850   0.031222  0.014433

the analysis explains that I can expect $0.27 dayly from AAPL, $3.71 monthly or $44.5199 yearly for any dollar we are investing.
#https://www.investopedia.com/terms/v/variance.asp#:~:text=The%20term%20variance%20refers%20to,by%20this%20symbol%3A%20%CF%832.
The term variance refers to a statistical measurement of the spread between numbers in a data set. More specifically, variance measures how far each number in the set is from the mean (average), and thus from every other number in the set. Variance is often depicted by this symbol: σ2. It is used by both analysts and traders to determine volatility and market security.
in the year AAPL shows 0.0186\% of volatility with 1.3649\% variation for this stock.
NVDA is more risky with 3.1222\% of variation in the price.

In [21]:
# mx portfolio monthly info
mx_portfolio_analysis_df['AAPL']['Month_outcome'] = us_portfolio_m_diff_df['AAPL'].mean()
mx_portfolio_analysis_df['AAPL']['Year_outcome']  = us_portfolio_m_diff_df['AAPL'].mean()*12
mx_portfolio_analysis_df['AAPL']['Year_var']      = us_portfolio_df["AAPL_Returns"].var()
#us_portfolio_analysis_df['AAPL']['Cov_year_mkt']  = us_portfolio_df["AAPL_Returns"].cov()
mx_portfolio_analysis_df['AAPL']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["AAPL_Returns"].var())
#amzn_df_m   = amzn_df.groupby(amzn_df.index.month).tail(1)
mx_portfolio_analysis_df['AMZN']['Month_outcome'] = us_portfolio_m_diff_df['AMZN'].mean()
mx_portfolio_analysis_df['AMZN']['Year_outcome']  = us_portfolio_m_diff_df['AMZN'].mean()*12
mx_portfolio_analysis_df['AMZN']['Year_var']      = us_portfolio_df["AMZN_Returns"].var()
#us_portfolio_analysis_df['AMZN']['Cov_year_mkt']  = us_portfolio_df["AMZN_Returns"].cov()
mx_portfolio_analysis_df['AMZN']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["AMZN_Returns"].var())
#fdx_df_m   = fdx_df.groupby(amzn_df.index.month).tail(1)
mx_portfolio_analysis_df['FDX']['Month_outcome'] = us_portfolio_m_diff_df['FDX'].mean()
mx_portfolio_analysis_df['FDX']['Year_outcome']  = us_portfolio_m_diff_df['FDX'].mean()*12
mx_portfolio_analysis_df['FDX']['Year_var']      = us_portfolio_df["FDX_Returns"].var()
#us_portfolio_analysis_df['FDX']['Cov_year_mkt']  = us_portfolio_df["FDX_Returns"].cov()
mx_portfolio_analysis_df['FDX']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["FDX_Returns"].var())

mx_portfolio_analysis_df

NameError: name 'mx_portfolio_analysis_df' is not defined